vm/src/stdlib/ctypes/basics.rs

```diff
@@ -0,0 +1,507 @@
+
+                    let offset_int = offset
+                        .into_option()
+                        .map_or(Ok(0), |off| i64::try_from_object(vm, off))?;
+
+                    if opts.readonly {
+                        Err(vm.new_type_error("underlying buffer is not writable".to_string()))
+                    } else if !opts.contiguous {
+                        Err(vm.new_type_error("underlying buffer is not C contiguous".to_string()))
+                    } else if offset_int < 0 {
+                        Err(vm.new_value_error("offset cannot be negative".to_string()))
+                    } else if cls_size > opts.len - (offset_int as usize) {
+                        Err(vm.new_value_error(format!(
+                            "Buffer size too small ({} instead of at least {} bytes)",
+                            cls_size,
+                            opts.len + (offset_int as usize)
+                        )))
+                    } else if let Some(mut buffer) = buffer.as_contiguous_mut() {
+                        // @TODO: Is this copying?
+
+                        let buffered = if copy {
+                            unsafe { slice::from_raw_parts_mut(buffer.as_mut_ptr(), buffer.len()) }
+                                .as_mut_ptr()
+                        } else {
+                            buffer.as_mut_ptr()
+                        };
+
+                        Ok(PyCData::new(
+                            None,
+                            Some(RawBuffer {
+                                inner: buffered,
+                                size: buffer.len(),
+                            }),
+                        ))
+                    } else {
+                        Err(vm.new_buffer_error("empty buffer".to_string()))
+                    }
+                }
+                Ok(_) => Err(vm.new_type_error("abstract class".to_string())),
+                Err(_) => {
+                    // @TODO: A sanity check
+                    Err(vm.new_type_error("attribute '__abstract__' must be bool".to_string()))
+                }
+            }
+        }
+        // @TODO: I think this is unreachable...
+        Err(_) => Err(vm.new_type_error("abstract class".to_string())),
+    }
+}
+
+pub fn default_from_param(
+    cls: PyTypeRef,
+    value: PyObjectRef,
+    vm: &VirtualMachine,
+) -> PyResult<PyObjectRef> {
+    //@TODO: check if this behaves like it should
+    if vm.isinstance(&value, &cls)? {
+        Ok(value)
+    } else if let Ok(parameter) = vm.get_attribute(value.clone(), "_as_parameter_") {
+        default_from_param(cls, parameter, vm)
+    } else {
+        Err(vm.new_attribute_error(format!(
+            "expected {} instance instead of {}",
+            cls.name,
+            value.class().name
+        )))
+    }
+}
+#[pyimpl]
+pub trait PyCDataFunctions: PyValue {
+    #[pymethod]
+    fn size_of_instances(zelf: PyRef<Self>, vm: &VirtualMachine) -> PyResult<PyObjectRef>;
```

Comment:
```suggestion
    fn size_of_instances(zelf: PyRef<Self>, vm: &VirtualMachine) -> PyResult;
```

`PyResult` == `PyResult<PyObjectRef>`